seatunnel-connectors-v2/connector-clickhouse/src/main/java/org/apache/seatunnel/connectors/seatunnel/clickhouse/sink/file/ClickhouseFileSinkWriter.java

```diff
@@ -105,56 +124,55 @@ private void nodePasswordCheck() {
     }
 
     @Override
-    public Optional<CKCommitInfo> prepareCommit() throws IOException {
-        return Optional.empty();
+    public Optional<CKFileCommitInfo> prepareCommit() throws IOException {
+        for (FileChannel channel : rowCache.values()) {
+            channel.close();
+        }
+        Map<Shard, List<String>> detachedFiles = new HashMap<>();
+        shardTempFile.forEach((shard, path) -> {
+            try {
+                List<String> clickhouseLocalFiles = generateClickhouseLocalFiles(path);
+                // move file to server
+                moveClickhouseLocalFileToServer(shard, clickhouseLocalFiles);
+                detachedFiles.put(shard, clickhouseLocalFiles);
+                // clear local file
+                clearLocalFileDirectory(clickhouseLocalFiles);
+            } catch (Exception e) {
+                throw new ClickhouseConnectorException(CommonErrorCode.FLUSH_DATA_FAILED, "Flush data into clickhouse file error", e);
+            }
+        });
+        rowCache.clear();
+        shardTempFile.clear();
+        return Optional.of(new CKFileCommitInfo(detachedFiles));
     }
 
     @Override
     public void abortPrepare() {
-
     }
 
     @Override
     public void close() throws IOException {
-        rowCache.forEach(this::flush);
+        for (FileChannel channel : rowCache.values()) {
+            channel.close();
+        }
     }
 
-    private void flush(Shard shard, List<SeaTunnelRow> rows) {
-        try {
-            // generate clickhouse local file
-            // TODO generate file by sub rows to save memory
-            List<String> clickhouseLocalFiles = generateClickhouseLocalFiles(rows);
-            // move file to server
-            attachClickhouseLocalFileToServer(shard, clickhouseLocalFiles);
-            // clear local file
-            clearLocalFileDirectory(clickhouseLocalFiles);
-        } catch (Exception e) {
-            throw new ClickhouseConnectorException(CommonErrorCode.FLUSH_DATA_FAILED, "Flush data into clickhouse file error", e);
-        }
+    private void saveDataToFile(FileChannel fileChannel, SeaTunnelRow row) throws IOException {
+        String data = this.readerOption.getFields().stream().map(field -> row.getField(this.readerOption.getSeaTunnelRowType().indexOf(field)).toString())
+            .collect(Collectors.joining("\t")) + "\n";
+        MappedByteBuffer buffer = fileChannel.map(FileChannel.MapMode.READ_WRITE, fileChannel.size(),
+            data.getBytes(StandardCharsets.UTF_8).length);
```

Comment:
Putting every row into `buffer` inside `mmap` region of file channel may cause the executor consumes too much RAM then crushes because of OOM.  In my opinion, Clickhouse File Connector is designed for huge data loading scenario where there will be multiples GB of data in every executor.And it is hard to put all these data in memory. 
Maybe we can make a tradeoff between speed and memory capacity.  We can dump some rows into disk when it accumulate to a certain threshold.

Reply:
Hi, can you tell me more about mmap consume too much RAM, because we map only one row each time. Anything I didn't know？

Reply:
I am not very sured about the behaviors of file channel. 
But when I using `fileChannel.map`, spark executors always crash. I currently use the following code as workaround.
```
ByteBuffer buffer = ByteBuffer.allocate(data.size());
buffer.put(out.toByteArray());
// flip offset
buffer.flip();
fileChannel.write(buffer);
```